---
layout: page
title: Reading
permalink: /reading/
---

Recommended reading prior to the workshop are in **bold**

### MSC Introduction
* Flouri T, Jiao X, Rannala B, Yang Z. 2018. Species tree inference with BPP using genomic sequences and the multispecies coalescent. Mol Biol Evol 35:2585-2593.
* Flouri T, Rannala B, Yang Z. 2020. A tutorial on the use of BPP for species tree estimation and species delimitation. Pp. 5.6.1-16 in Scornavacca C, Delsuc F, and Galtier N, eds. Phylogenetics in the Genomic Era. No Commercial Publisher.

### Introduction to Bayesian Phylogenetics and Gene Flow
* **Nascimento FF, dos Reis M, Yang Z. 2017. A biologist's guide to Bayesian phylogenetic analysis. Nature Ecol Evol 1:1446-1454.**
* **Hibbins MS, Hahn MW. 2022. Phylogenomic approaches to detecting and characterizing introgression. Genetics 220:10.1093/genetics/iyab1173.**

### Bayesian MSC and Gene Flow Models
* Flouri T, Jiao X, Rannala B, Yang Z. 2020. A Bayesian implementation of the multispecies coalescent model with introgression for phylogenomic analysis. Mol Biol Evol 37:1211-1223.
* Flouri T, Jiao X, Huang J, Rannala B, Yang Z. 2023. Efficient Bayesian inference under the multispecies coalescent with migration. Proc Nat Acad Sci USA:in press, 10.1073/pnas.2310708120.
* Ji J, Jackson DJ, Leache AD, Yang Z. 2023. Power of Bayesian and heuristic tests to detect cross-species introgression with reference to gene flow in the Tamias quadrivittatus group of North American chipmunks. Syst Biol. 72:446-465.
* **Jiao X, Flouri T, Yang Z. 2021. Multispecies coalescent and its applications to infer species phylogenies and cross-species gene flow. Nat Sci Rev 8:DOI: 10.1093/nsr/nwab1127.**
* Tiley GP, Flouri T, Jiao X, Poelstra JW, Xu B, Zhu T, Rannala B, Yoder AD, Yang Z. 2023. Estimation of species divergence times in presence of cross-species gene flow. Syst Biol. 72:820-836.

### Network Inference from Quartets
* Solís-Lemus C, Ané C. 2016. Inferring phylogenetic networks with maximum pseudolikelihood under incomplete lineage sorting. PLoS Genetics. 12:e1005896.
* Yu Y, Nakhleh L. 2015. A maximum pseudo-likelihood approach for phylogenetic networks. BMC Genomics. 16:S10

### Empirical Examples
* Crowl AA, Fritsch PW, Tiley GP, Lynch NP, Ranney TG, Ashrafi H, Manos PS. 2022. A first complete phylogenomic hypothesis for diploid blueberries (Vaccinium section Cyanocossus). Am J Bot. 109:1596-1606.
* Karimi N, Grover CE, Gallagher JP, Wendel JF, Ané C, Baum DA. 2020. Reticulate evolution helps explain apparent homoplasy in floral biology and pollination in baobabs (Adansonia; Bombacoideae; Malvaceae). Syst Biol. 69:462-478.
* Morales-Briones DF, Kadereit G, Tefarikis DT, Moore MJ, Smith SA, Brockington SF, Timoneda A, Yim WC, Cushman JC, Yang Y. 2021. Disentangling sources of gene tree discordance in phylogenomic data sets: testing ancient hybridizations in Amaranthaceae s.l. Syst Biol. 70:219-235.
* Thawornwattana Y, Seixas FA, Mallet J, Yang Z. 2022. Full-likelihood genomic analysis clarifies a complex history of species divergence and introgression: the example of the erato-sara group of Heliconius butterflies. Syst Biol 71:1159-1177.